### ISSUE TITLE
Cannot get instance of ActionRouter in ACTION_ACT_PREPROCESS

### ISSUE THREAD

**Issue by michitux** [splitbrain/dokuwiki#2373]
As `ACTION_ACT_PREPROCESS` is triggered from within the constructor of the `ActionRouter`, the `$instance` class variable hasn't been assigned yet. Therefore, calling `ActionRouter::getInstance()` from within `ACTION_ACT_PREPROCESS` instantiates a new `ActionRouter` instance in which `ACTION_ACT_PREPROCESS` is triggered again which creates an endless loop. Currently I see no way to interrupt this loop without triggering user-visible messages. I guess a proper fix would be to trigger action processing in a separate method that is not called in the constructor.

I am currently trying to fix the [saveandedit plugin](https://www.dokuwiki.org/plugin:saveandedit). My current attempt is to get an instance of `ActionRouter` in `ACTION_ACT_PREPROCES` as I want to execute the `save` action from within `ACTION_ACT_PREPROCESS` in order to trigger the `edit` action afterwards (intercepting the `ActionAbort` exception for `draftdel` and replacing it by an `ActionException` with `edit`). Do you have any other ideas for implementing this? Basically I want to execute the `save` action with all checks, do some cleanup and then let DokuWiki continue with a regular `edit` action.

**Comment by michitux** [comment]
Okay, I found a possible workaround: one can get the action router object from `debug_backtrace()`. Further, I am wondering if it would be cleaner to intercept the `draftdel` action directly (in contrast to previous releases, the event is triggered again) and to change the action to edit there (if the previous action was `save`). So in other words: there is a workaround, and it is possibly not even needed.

However, I still think this should be fixed.

**Comment by splitbrain** [comment]
Maybe we could pass a reference to the action router ($this) in the event data? Would that help?

**Comment by micgro42** [comment]
Yes, that problem came up before as well. I started to look into fixing it some time ago, but it never let anywhere, though I don't remember why.  See b384ead49ef969930ad207f188322aaea30982f9

**Comment by michitux** [comment]
@micgro42 This is exactly the fix I had in mind. A much shorter fix for the singleton pattern would be to add `self::$instance = $this;` at the beginning of the constructor (but this doesn't fix the issue you wanted to fix I think).

@splitbrain While passing a reference in the event would fix that particular problem, I don't see any reason for this if the action router is supposed to be a singleton. 

I found a completely different (and imho better) fix for the saveandedit plugin (see [action.php#l66ff](https://github.com/michitux/dokuwiki-plugin-saveandedit/blob/8abd09b7f166552d107fe02d92402bbd47140f22/action.php#L66)) where I don't need to use the action router explicitly. I'm still not completely sure if this is the best fix though, see michitux/dokuwiki-plugin-saveandedit#7.
